### PR TITLE
Add Llamacpp support

### PR DESCRIPTION
**Pull request by FredericoPerimLopes**
Add Llamacpp support

### CODE REVIEW

File: crates/llm-ls/src/backend.rs [hunk 1]
```diff
@@ -172,6 +203,15 @@ pub(crate) fn build_body(
             request_body.insert("model".to_owned(), Value::String(model));
             request_body.insert("stream".to_owned(), Value::Bool(false));
         }
+        Backend::LlamaCpp { .. } => {
+            if model.to_lowercase().contains("deepseek") {
+                request_body.insert("prompt".to_owned(), Value::String(format!("\"{}\"", prompt)));
+            } else {
```
Comment (McPatate) [inline crates/llm-ls/src/backend.rs:209]: Why is this necessary?

File: crates/llm-ls/src/error.rs [hunk 2]
```diff
@@ -31,6 +31,8 @@ pub enum Error {
     InvalidTokenizerPath,
     #[error("ollama error: {0}")]
     Ollama(crate::APIError),
+     #[error("llamacpp error: {0}")]
+    LlamaCpp(crate::APIError),
```
Comment (McPatate) [inline crates/llm-ls/src/error.rs:35]:
```suggestion
    #[error("llamacpp error: {0}")]
    LlamaCpp(crate::APIError),
```

File: crates/custom-types/src/llm_ls.rs [hunk 3]
```diff
@@ -67,10 +67,10 @@ pub enum Backend {
         #[serde(default = "hf_default_url", deserialize_with = "parse_url")]
         url: String,
     },
-    // TODO:
-    // LlamaCpp {
-    //   url: String,
-    // },
+
```
Comment (McPatate) [inline crates/custom-types/src/llm_ls.rs:70]:
```suggestion
```

**Review by McPatate — COMMENTED** [review]
Could you run rustfmt before comitting please?

File: crates/llm-ls/src/backend.rs (hunk 1 above)
Comment (FredericoPerimLopes) [reply crates/llm-ls/src/backend.rs:209]: @McPatate When I was doing tests I used deepseek as the base model. Its payload required the prompt to be between quotes, so that's the reason.

**Comment by McPatate** [comment]
Hi @FredericoPerimLopes, sorry for the late reply. I'm not convinced of the need for this PR as LlamaCpp supports OpenAI-style API.

Is there value in supporting the "native" llamacpp API?

**Comment by rggs** [comment]
Ok here's why I think this is necessary.
1. llama.cpp implements the [OpenAI chat api](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md), not the [legacy completions api](https://platform.openai.com/docs/api-reference/completions/create).
2. llama.cpp can, however, accept a request in the form of the legacy completion api via their `/completion` endpoint (as opposed to `/v1/completions` for the OAI models).
3. The legacy AOI [completions api repsonse](https://platform.openai.com/docs/api-reference/completions/object) is of the form `response.choices`, so you get the first item of the choices array.
4. The [llama.cpp completion response](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#result-json) is of the form `response.content`.

So it seems like the easiest option is to either create a separate llama.cpp backend adapter, or create an additional config option for something like `llama.cpp_response = True` so that it uses the OpenAI adapter, but gets the completion from `response.content` instead of `response.choices`.

The end result of all this is I'm able to make a request to the llama.cpp server but none of the suggestion shows in VSCode.

**Comment by julien-c** [comment]
would be cool imo!

**Comment by McPatate** [comment]
Closing in favor of https://github.com/huggingface/llm-ls/pull/94

**Comment by McPatate** [comment]
@FredericoPerimLopes thank you for your contribution, I added you as co-author on the commit that landed in main
